front/src/common/Map/Search/MapSearchStation.jsx

```diff
@@ -4,69 +4,83 @@ import { useSelector, useDispatch } from 'react-redux';
   const infraID = useSelector(getInfraID);
 
+  const [postSearch] = osrdEditoastApi.usePostSearchMutation();
   const dispatch = useDispatch();
 
   const { t } = useTranslation(['map-search']);
 
-  const updateSearch = async (params) => {
-    try {
-      const data = await post(SEARCH_URL, params);
-      setSearchResults(data);
-    } catch (e) {
-      /* empty */
-    }
+  const updateSearch = async (params, setResults) => {
+    const { data, error } = await postSearch({
+      body: params,
+    });
+    if (error) console.error(error);
+    setResults(data);
   };
 
   const debouncedSearchTerm = useDebounce(searchState, 300);
 
-  const getPayload = () => {
-    let trigram = null;
-    let name = null;
-    if (!dontSearch && debouncedSearchTerm) {
-      if (searchState.length < 3) {
-        trigram = searchState;
-        name = null;
-      } else if (searchState.length === 3) {
-        trigram = searchState;
-        name = searchState;
-      } else if (searchState.length > 3) {
-        name = searchState;
-        trigram = null;
-      }
-    }
+  // Create playload based on the type of search "name" or "trigram"
+
+  const createPlayload = (queryType) => {
+    const searchQuery =
+      queryType === 'name' ? ['search', ['name'], searchState] : ['=i', ['trigram'], searchState];
     const payload = {
       object: 'operationalpoint',
-      query: [
-        'and',
-        ['or', ['search', ['name'], name], ['search', ['trigram'], trigram]],
-        ['=', ['infra_id'], infraID],
-      ],
+      query: ['and', searchQuery, ['=', ['infra_id'], infraID]],
     };
     return payload;
   };
 
+  // Sort on name, and on yardname
+  const orderResults = (results) =>
+    [...results].sort((a, b) => a.name.localeCompare(b.name) || a.ch.localeCompare(b.ch));
+
+  const getResult = async () => {
+    if (!searchResults || searchResults.length > 3) setTrigramResults([]);
+
+    if (searchState.length < 3) {
+      setNameResults([]);
+      const searchTrigramQuery = createPlayload('trigram');
+      updateSearch(searchTrigramQuery, setTrigramResults);
+    } else if (searchState.length === 3) {
+      // The trigram search should always appear first, we need two api calls here.
+      const searchTrigramQuery = createPlayload('trigram');
+      updateSearch(searchTrigramQuery, setTrigramResults);
+      const searchNameQuery = createPlayload('name');
+      updateSearch(searchNameQuery, setNameResults);
+    } else if (searchState.length > 3) {
+      const searchNameQuery = createPlayload('name');
+      updateSearch(searchNameQuery, setNameResults);
+    }
+  };
+
   useEffect(() => {
-    if (searchState) updateSearch(getPayload());
+    getResult();
+
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [debouncedSearchTerm]);
 
+  useEffect(() => {
+    setSearchResults([...trigramResults, ...orderResults(nameResults)]);
+  }, [trigramResults, nameResults]);
```

Comment:
You can directly do this (in searchNames)
```
setNameResults(orderResults(nameResults))
```
And then
```suggestion
    setSearchResults([...trigramResults, ...nameResults]);
  }, [trigramResults, nameResults]);
```

Reply:
Done !